### ISSUE TITLE
Recurrent Dropout is Wrong

### ISSUE THREAD

**Issue by OverLordGoldDragon** [tensorflow/tensorflow#35724]
I've reviewed one design in-depth, and two others superficially, but Keras/TF's `recurrent_dropout` does not implement _any_ of them; publication links below.

 1. I see some potentially severe problems with TF's implementation in light of the papers I've read, which explicitly advocate against the used scheme. This said - what is TensorFlow / Keras's justification / rationale  of its own implementation?

 2. The implementation is inconsistent - see below; the docstring only mentions a performance difference, but there's also a _reproducibility_ and _design_ difference; `==1` uses _different masks_ per gate, whereas `==2` uses a _shared mask_. The two are neither theoretically nor practically identical. 

Second's fixable via a docstring, but first involves significant changes to recurrent dropout logic for `LSTM`, `GRU`, and maybe other RNNs. This said: **is TensorFlow / Keras open to changing its base implementations of recurrent dropout?** If so, I can go ahead and clarify **(1)** in detail, and maybe even do the re-implementing myself in a PR, per paper 1.

<hr>

**Inconsistency**: `implementation==1` vs. `implementation==2`

```python
if 0. < self.recurrent_dropout < 1.:  # implementation==1
    h_tm1_i = h_tm1 * rec_dp_mask[0]
    h_tm1_f = h_tm1 * rec_dp_mask[1]
    h_tm1_c = h_tm1 * rec_dp_mask[2]
    h_tm1_o = h_tm1 * rec_dp_mask[3]
```
```python
if 0. < self.recurrent_dropout < 1.:  # implementation==2
    h_tm1 *= rec_dp_mask[0]
```
Source codes: [keras](https://github.com/keras-team/keras/blob/master/keras/layers/recurrent.py#L2014) -- [tf.keras](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/recurrent.py#L2391)

<hr>

**Publications**:

 1. [Recurrent Dropout without Memory Loss](https://arxiv.org/abs/1603.05118)
 2. [A Theoretically Grounded Application of Dropout in Recurrent Neural Networks](https://arxiv.org/abs/1512.05287)
 3. [RNNDrop: A novel dropout for RNNs](https://sci-hub.se/10.1109/ASRU.2015.7404775)

**Comment by qlzh727** [comment]
Hi @OverLordGoldDragon. Thanks for reporting the issue.

1. For point 1, can you be more specific about the difference between the paper and the keras implementation? eg the math formula in the paper vs the code in keras.

2. For point 2, I have previously also notice the difference, and added a patch to force use implementation 1 when the recurrent_dropout is enabled. The code is in 9002ff156ef173e91822a9ad9ba51f260e4a2223.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 2: I see, so TF's ahead of Keras in this; nicely done.

1: sure. Upon closer inspection of the second paper, I realize TF's implementation is probably based on it - but is still off:

 1. `tf.nn.dropout` is _inverted dropout_, whereas the authors don't apply any scaling; their code on the paper's last page generates masks via [`srng.binomial()`](https://github.com/Theano/Theano/blob/master/theano/sandbox/rng_mrg.py#L896) - no scaling here, nor in the formulas.
 2. The paper applies a _single_ dropout mask for the entire sequence; I couldn't really determine from the code whether this is the case, or masks are generated per-timestep. I'm fairly sure, however, that a new set of masks is generated with a new input to the model, _even with_ `stateful=True`.

Implications:

 1. **Dampened gradient**: upscaling preactivations drives the recurrent sigmoid into saturation sooner, exacerbating gradient vanishing.
 2. **Stateful** is mainly for fitting a _single sequence_ in separate consecutive fits; that is, inputs are assumed to _causally continue_ from their predecessors. The paper's central design element is dropping the _same features per_ (entire) _sequence_ (see f1).

There may be more as I haven't read the paper in full, but the idea is, TF's implementation concretely differs from the paper's in at least these two regards.

<hr>

**Footnotes**:  f1 - pg. 6: 

> we drop the same words throughout the sequence - i.e. we drop word types at random rather than word tokens (e.g. the sentence "the dog and the cat" might become "__ dog and __ cat" or "the __ and the cat", but never "__ dog and the cat")

In context of signals, this'd mean dropping the same _channels_ for the entire sequence. (In my application, the signal is 240,000 timesteps per sequence, requiring stateful even post dim-reduction)

**Comment by OverLordGoldDragon** [comment]
@qlzh727 More importantly, however, I find paper 1's implementation more viable than 2's from my reviews so far - and 1 does present empirically better-or-equal results than 2. It seems to offer superior gradient dynamics for long timeseries, _and_ is `implementation=2`-friendly. Further, 2 accounts for train vs. interence-time scaling, whereas 1 seems to not use any scaling (which may not be best). 

There's more to be argued in favor of paper 1, but before I do I should ask - **is TensorFlow open to changing its base implementation of recurrent dropout, OR enabling a settable alternative?** I.e. e.g. `recurrent_dropout_implementation=2`. I can see it implemented with a relatively small change to the overall source code.

I'm aware the answer to above is a "yes" if offering it as a "custom add-on" - hence my asking on _base_. If TF is open to this end, I can detail further as to why 1 is better than 2, or is as good as, and hence should be in the base implementation.

**Comment by qlzh727** [comment]
Thanks for the details.

1. We are generally open for change of the layer behavior if there is a mis-interpretation for the concept in the paper.

2. For those 2 paper referred, I can see first paper is quite clear that not scaling up when dropout is applied (sample code at last page). On the other hand, the second paper is explicitly scaling up the value when dropout (formula 18). It's unclear to me that which one is the "correct" approach. If one approach is generically better than the other, then we should change to that, otherwise, we have to allow user control this behavior.

3. Reusing the same mask for the whole sequence. Keras is using same mask for the whole sequence (variational dropout). The mask will be reset for every batch, but stays the same for all the timesteps within the batch. You can find more details in 9650c977e3b3ca73a0249a08f703960f478b3ea9.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 Thanks for the feedback; glad to hear TF is open to change with sufficient reason. 

 2. Second paper max formula number is 9, pg. 5; which is "formula 18"? 

 3. Right, so then it _changes_ between batches even with stateful, which may be undesirable per paper 2

I have more to say, but will relay it after clarification; awaiting your response.

**Comment by qlzh727** [comment]
Sorry, I think I made the typo between paper 1 and 2 in my previous message. I was referring to paper "Recurrent Dropout without Memory Loss", on page 4,  formula 18, with content
```
mask ∗ x/p, if train phase
x otherwise
```
Note here p is the probably for dropout, dividing p means scale up for the value.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 Indeed, it is from paper 1, whereas Keras' implementation is based on _2_. As noted in my previous reply, 2's source code at the last page uses no scaling, as confirmed with a lookup to [`srng.binomial()`](https://github.com/Theano/Theano/blob/master/theano/sandbox/rng_mrg.py#L896). i.e., Keras' implementation _conradicts the authors'_.

Regarding formula 18, authors mention it as an example of a _problem_ they're fixing right underneath:

> In this case the above argument holds as well, but instead of observing exponentially decreasing hidden states during testing, we will observe exponentially increasing values of hidden states during
training.

 > Our approach **addresses the problem** discussed previously by ...

From the source code linked in the paper, we can see that [no scaling](https://github.com/stas-semeniuta/drop-rnn/blob/master/layers.py#L99) occurs on train time, but it does at [test time](https://github.com/stas-semeniuta/drop-rnn/blob/master/layers.py#L143) (`drop_candidates` is for 2's implementation, which they compare against paper 3's implementation via `=False`).

**Comment by qlzh727** [comment]
Thank you for the very detailed explanation and sorry for the late reply.

1. I agree with you that the recurrent dropout should not scale up the undroped value, wrt to all the referred paper.
2. Wrt to stateful RNN, currently generate new masks for each of the mini batch, but we can update that to handle the stateful case. Within each of the batch, the masks are the same.
3. If we going to change the recurrent dropout to be not scaling, probably we will add a new flag for this since its changing the existing behavior. Even through it might not be correct in the first place, but people relied on it.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 All of that sounds good, and I agree with 3 also. Further, I've changed my position since last time: I no longer advocate for _replacing_ current recurrent dropout, as I just achieved remarkable performance with current dropout, which wouldn't be possible if it did more harm than good. 1's authors do also note that performance was close to 2's on multiple benchmarks.

This said, and my case considered, is TF open to a PR on paper 1's implementation of `recurrent_dropout`? (e.g. via `recurrent_dropout_implementation=2`)

**Comment by qlzh727** [comment]
Yes, I think we are open for that. Also adding @fchollet  as the API owner to confirm this.

We can have a more descriptive naming for the new param, but we can decide it later.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 Wonderful. To clarify, did fchollet confirm it, or are you requesting a confirmation?

If it's a green light, I'll go ahead and cook up a PR - though unsure when exactly, I'd say within a few weeks. As for correcting the current implementation, I'll leave that to the devs.

**Comment by qlzh727** [comment]
@OverLordGoldDragon, I would like to have @fchollet to confirm this is ok to proceed as the API owner.

**Comment by OverLordGoldDragon** [comment]
I confirmed numerous times again, that `recurrent_dropout` promotes exploding gradients (`nan` loss & weights) - but this time, with an insight: `recurrent_dropout` + _large weight norms_, or `recurrent_dropout` + _uncentered inputs_ (e.g. ReLU). While both can cause divergence without `recurrent_dropout`, `recurrent_dropout` significantly lowers the required threshold.

To test my hypothesis on scaling, edited `_recurrent_mask *= (1 - self.recurrent_dropout)` - and indeed, explosions no more, and training proceeds in a stable manner.

**Comment by MariosKef** [comment]
Has this been fixed by any chance in version 2.4.1? 
Thanks!

**Comment by OverLordGoldDragon** [comment]
@marioskef TensorFlow 4 will happen sooner.

I'd look for third-party implementations.

**Comment by MariosKef** [comment]
@OverLordGoldDragon thanks for your reply! 
I am asking because here https://github.com/yaringal/BayesianRNN/issues/5#issuecomment-354255980 the author of the _A Theoretically Grounded Application of Dropout in Recurrent Neural Networks_ said he implemented it for keras. So I am a bit confused tbh what is what.

**Comment by OverLordGoldDragon** [comment]
@MariosKef Don't recall the details, but these bits seem relevant (Ctrl + F):

> Upon closer inspection of the second paper, I realize TF's implementation is probably based on it - but is still off:

> More importantly, I find paper 1's implementation more viable than 2's from my reviews so far

So I've concluded it's not based on paper 2 (you could ask the author directly, point to this thread), and even if it was, it's not the best approach.

**Comment by MariosKef** [comment]
@OverLordGoldDragon 

Thanks for your answer. So basically what is currently implemented in TF is a type of variational dropout but not the one suggested from the paper, if got it correctly.

Any advice for 3rd implementations or other libraries?

**Comment by OverLordGoldDragon** [comment]
@MariosKef Right, it _partly_ implements 2. I don't know of such libraries.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
